lib/elixir/src/elixir_expand.erl

```diff
@@ -696,18 +696,36 @@ assert_no_ambiguous_op(Name, Meta, [Arg], E) ->
 assert_no_ambiguous_op(_Atom, _Meta, _Args, _E) ->
   ok.
 
+assert_no_clauses(_Name, _Meta, [], _E) ->
+  ok;
+assert_no_clauses(Name, Meta, Args, E) ->
+  assert_arg_with_no_clauses(Name, Meta, lists:last(Args), E).
+
+assert_arg_with_no_clauses(_Name, _Meta, Arg, _E) when not is_list(Arg) ->
+  ok;
+assert_arg_with_no_clauses(_Name, _Meta, [], _E) ->
+  ok;
+assert_arg_with_no_clauses(Name, Meta, [{_, [{'->', _, _} | _]}], E) ->
```

Comment:
We are not checking this case:

```elixir
wrong_fun do
  foo -> bar
after
  :ok
end
```

Because this clause is matching exclusively on the last entry as it doesn't consider the tail. Given @michalmuskala feedback, I think replacing all clauses above by these two would be fast and consider all cases:

```erlang
assert_arg_with_no_clauses(Name, Meta, [{Key, Value} | Rest], E) when is_atom(Key) ->
  case Value of
    [{'->', _, _} | _] ->
      form_error(Meta, ?key(E, file), ?MODULE, {invalid_clauses, Name});
    _ ->
      assert_arg_with_no_clauses(Name, Meta, Rest, E)
  end;
assert_arg_with_no_clauses(_Name, _Meta, _Arg, _E) ->
  ok.
```